src/lib.rs

```diff
@@ -146,151 +148,166 @@ impl<Types: IpfsTypes> Ipfs<Types> {
         let dag = IpldDag::new(repo.clone());
         let ipns = Ipns::new(repo.clone());
 
-        Ipfs {
+        UninitializedIpfs {
             repo,
             dag,
             ipns,
-            repo_events: Some(repo_events),
-            swarm: Some(swarm),
+            moved_on_init: Some((repo_events, swarm)),
             exit_events: Vec::default(),
         }
     }
 
     /// Initialize the ipfs repo.
-    pub fn init_repo(&self) -> impl Future<Output=Result<(), Error>> {
-        self.repo.init()
+    pub async fn start(mut self) -> Result<(Ipfs<Types>, impl std::future::Future<Output = ()>), Error> {
+        use futures::compat::Stream01CompatExt;
+
+        let (repo_events, swarm) = self.moved_on_init
+            .take()
+            .expect("Cant see how this should happen");
+
+        self.repo.init().await?;
+        self.repo.init().await?;
+
+        let (sender, receiver) = channel::<IpfsEvent>(1);
+        self.exit_events.push(sender);
+
+        let fut = IpfsFuture {
+            repo_events,
+            exit_events: receiver,
+            swarm: swarm.compat(),
+        };
+
+        let UninitializedIpfs { repo, dag, ipns, exit_events, .. } = self;
+
+        Ok((Ipfs {
+            repo,
+            dag,
+            ipns,
+            exit_events
+        }, fut))
     }
+}
 
-    /// Open the ipfs repo.
-    pub fn open_repo(&self) -> impl Future<Output=Result<(), Error>> {
-        self.repo.open()
+impl<Types: IpfsTypes> Ipfs<Types> {
+    /// Creates a new ipfs node.
+    pub fn new(options: IpfsOptions<Types>) -> UninitializedIpfs<Types> {
+        UninitializedIpfs::new(options)
     }
```

Comment:
This change is quite questionable ... Not sure if it was needed at all; all of the init work could be done inside `Ipfs::new` but it would need to be async. I ended up with this solution while investigating the `Sync` bounds issue.